### PR TITLE
Revised Status Legend popup

### PR DESCRIPTION
**Pull request by bergware**
- Legend is moved to a fixed position under "Device" or "Name"
- Improved description and working of device spin up/down controls

### CODE REVIEW

**Comment by limetech** [comment]
Actually I think it's better to not have this pop-up at all, instead put it in the 'Help' for the page.  This makes it more consistent.

I like the idea of rollover text giving description of status represented at that moment by the icon.

**Comment by limetech** [comment]
For the icons, let's try this:

green-on, green-blink = same as it is now (orb)
green-off = unneeded, never have this state

yellow-on, yellow-blink = same as it now (triangle)
yellow-off = unneeded, never have this state

red-on = lets go with bright red "X"
such as http://icons.iconarchive.com/icons/kyo-tux/phuzion/48/Sign-Error-icon.png
red-blink = same as red-on.  This is a state we should never see because once a disk is disabled, unraid will never access it for any purpose, not even to spin it down.  Hence drive should always be spun up.  We don't access it because depending on nature of fault, could cause 'hangs' because of long timeouts.
red-off = same as red-on.  This is because user will see that disk is not installed because it will say "not installed" in ID column.  The red X is big indicator they still have a disabled disk.

blue-on, blue-blink = let's go with a blue/dim blue square
blue-off = not needed

grey-off = same as current: grey square.  This probably looks the same as "blue-blink", but that's ok because it will say "Not installed" in the Identification column.

As for size, let's keep it the same.  For people who think it's too small they can zoom their browser.

**Comment by limetech** [comment]
Off topic: as long as you're "in the code" :) ....

I like the naming of the diagnostics file:
<server-name>-diagnostics-<time>

I'd like to see similar treatment for syslog:
<server-name>-syslog-<time>

and smart info:
<server-name>-<disk-name>-smart-<time>

(<disk-name> is parity disk1 disk2 .. cache cache2 ..)

Also maybe move that "Download" button to the bottom of the Self-tests section?  Maybe call it "Download SMART"?

**Comment by bergware** [comment]
Just did a commit to move the status indicator information to the help section, which is indeed more logical (it will be the first place for people to look).

The icons on the main page now have a state indicator when hovering over them.

Perhaps you want to have a look at the various descriptions and correct/improve where needed.

**Comment by bergware** [comment]
About the icons:

1. A disabled array will also show a red X, consequently, is this what you want ?
2. We should change the state for cache only user shares (red X), can become blue-on ?

Anyway I make the changes so they can be reviewed, will give a better impression of how things look.

**Comment by bergware** [comment]
To view all status indicators -turn on 'Help' and visit the following pages

main - array devices (ArrayDevices.page)
main - cache devices (CacheDevices.page)
shares - user shares  (ShareList.page)
shares - disk shares (DiskList.page)

Note: status indicator on main page is created in /include/DeviceList.php

Please adapt text were needed/appropriate. Sure your wording is better than mine!

**Comment by limetech** [comment]
Hey this turned out nice, right?  Excellent work!

Was about to release -rc4, but now I'm going to integrate this and change just a little bit of the rollover wording.

For shares I'm just going to have two states:
* green-on = all share data is on valid parity-protected array and/or valid cache pool (of 2 or more devices)
* yellow-on = all or some share data is "at risk" of loss if the wrong device fails

In other words, I will make the 'color' indicate if underlying storage of share is redundant or not.  For example, if someone creates an array with no parity disk, then all disk shares will be yellow-on, and all user shares with any files on array will be yellow-on.

What is the 'User Utilities' for?

**Comment by bergware** [comment]
"User Uitilities" is the section under "Settings" where user made plugins can put their settings icon.

I've found that most plugin authors put their settings under "User Utilities", but don't realize it isn't standard, which means that if they don't include the page and it isn't already created by another plugin then nothing is shown.

Including it as standard will make it easier for developpers.

Ps. When you change the share states, then do you also update the help text ?

**Comment by bergware** [comment]
And yes, I do like the icons after your recommendations.

**Comment by limetech** [comment]
Ok good idea to include Uer Utilities page.  Maybe when we implement your package manager (where limetech maintains set of packages that all plugins use - nice idea btw), we can put a tool there in position 1 that will show the list of downloaded packages.  I hope to implement this after 6.0 'stable' is released, that is in 6.1-beta1.

Yes hopefully I don't forget to change help text too :)

Another off-topic question: how would we go about integrating a jquery file upload utility?  That is, I'd like to be able to upload files from user's PC to the server:
* To upload a thumbnail image or avatar for users.
* To upload a custom banner image.
* To upload a key file.
* To upload a release zip file.

**Comment by limetech** [comment]
Also, I was thinking to get rid of "Uncommitted changes protection" confirmation setting (have it behave as if always set to 'Yes').  Agreed?

**Comment by bergware** [comment]
For the moment I would leave it as is, it doesn't really hurt, and consider this for 6.1 together with a possible simplification of the display settings (don't make some settings optional anymore). The shear number of display settings may be quite daunting for some.

**Comment by bergware** [comment]
With your permission I like to keep involved in further development of the webGUI beyond 6.0. So any any plans for the next version, I am happy to participate.

**Comment by limetech** [comment]
> For the moment I would leave it as is

The reason I wanted to force it to 'Yes' was that I added the checks in NetworkSettings.page to disable all the input fields if array is Started (changing network settings while docker/vm's are running makes them freak out).  But I didn't put the check on the Apply button because with 'Uncommited changes protection' enabled the Apply button only gets enabled when a field is changed, and hence if you can't change a field Apply never gets enabled.  But we can do this later, I'll just add the 'disabled' check to the Apply button there.

**Comment by limetech** [comment]
> With your permission I like to keep involved in further development of the webGUI beyond 6.0. So any any plans for the next version, I am happy to participate.

I'm hoping you would continue to participate!  I'm learning a lot from you.

**Comment by bergware** [comment]
Another 'change' for your consideration...

Actually I don't know if this going to cause more confusion, but here it goes!

With the introduction of a two state (green/yellow) state for user shares, it might make sense to use the same states for the disk shares. If the array is protected then all disk shares are green, if the array is unprotected then the disk shares are yellow.

Now the confusing part may come from the same names "disk X"  used on different pages and in different context, the user may get confused to see state A on the main page while the shares page shows state B.

What do you think?

**Comment by limetech** [comment]
Yes that was what I meant above by "disk shares being at risk".  Also, should change the entries in the Name column to their proper identifier as seen on the network, e.g.,

"Disk 1" => "disk1"
"Cache" => "cache"

That will probably help.  (Also make same change on SecuritySMB/AFP/NFS.page files for disk shares).

Also there's a bug: if there is a cache pool (2 or more devices) there should only be one line in that table for 'cache' share, even if 'Cache' device does not exist.

**Comment by limetech** [comment]
BTW, introducing disk['fsColor'] which will be:
grey-off => not mounted
green-on => mounted, underlying device has redundancy/protection
yellow-on => mounted, underlying device does not have redundancy/protection

**Comment by bergware** [comment]
Changing "Disk 1" into "disk1", etc, is easy ... done !

I use the following code to skip entries:

`if ($name=='parity' || $name=='flash' || preg_match('/cache[0-9]+/',$name) || $disk['status']=='DISK_NP') continue;`

So it should skip names like 'cache2, cache3, etc but not 'cache'.

You are saying name 'cache' may not exist in the array yet cacheX is there ?

Will anticipate on disk['fsColor'] but can't really test!

**Comment by limetech** [comment]
Right, once disk['fsColor'] is there you only need to test that it's not "grey-off".

When someone has a single-device cache disk (number of cache slots == 1), everything works the same as for disk shares, in fact, cache is just another disk share.

When there is a cache pool (number of cache slots > 1), then things are a little different:
* On Main page the 'cache' link is always active even if no physical device is present in that slot.  That is the purpose of that 'DISK_OK_NP" status.  You click this link to do all pool operations such as scrub and balance.
* On Main page, the 'cache2', 'cache3',.., links just give basic info about the device itself.  You cannot initiate scrub/balance from these pages.
* On Shares page, if cache is mounted, then only line that gets displayed is the 'cache' line, again, no matter if physical device is present in that slot or not.  In this case 'cache' refers to a multi-device pool.

When there is no device plugged into 'cache' slot but there is in 'cache2' slot, then it's not very obvious to the use that pool operations still take place via 'cache' link (because the link itself is not very obvious).  I've been thinking about this and may end up letting user initiate all btrfs pool operations via any of the member devices - that's how the btrfs utils themselves work too.

We plan on letting user create multiple "pools" in a future release.

**Comment by bergware** [comment]
Yeah, I am not using a cache pool myself and don't have the possibility at the moment to set up one.

So it is a bit driving in the dark for me !

If I understand your story correctly then replacing my previous multiple test with

`if ($disk['fsColor'] == 'grey-off') continue`

Will do all the magic, or would this still display multiple cache shares?

**Comment by bergware** [comment]
Another question ...

Shouldn't we add a condition to the SMART attributes/Capabilities/Ident pages to test on disk status 'NP', otherwise we run into displaying smart information for a non-physical disk ?

**Comment by limetech** [comment]
Yes the disk fsColor check should do the trick!

Right about the SMART for _NP status - that's a bug I saw and forgot to write down...

**Comment by bergware** [comment]
And that makes another pull request, see github !!!!
